<template> layering proposal (#813)
emberjs/rfcs: text/0813-template-layering.md

export const Foo = template`<Hello />`;

// export class Bar {
//   <template><Hello /></template>
// }
export class Bar {
  static {
    template`<Hello />`;
  }
}
```

In addition, we also propose a new `@ember/template-compilation/runtime`
module as an opt-in to enable runtime template compilation.

## Motivation

[RFC #779][rfc-779] introduced a first-class component syntax feature that has
numerous benefits, among which are:

1. Opt-in to strict mode ([RFC #496][rfc-496])
2. Eliminating the need for name-based runtime resolutions
3. Ability to use imported template constructs
4. Ability to interact with the surrounding JavaScript scope

[rfc-496]: ./0496-handlebars-strict-mode.md

However, one of the drawbacks is that it requires a custom syntax extension
(`<template>`) and a custom file format (`.gjs/.gts`), which requires a build
step and other custom tooling.

RFC #779 makes a good case for why this is necessary and preferable to the
alternatives. We stand by the rationales given in RFC #779, and this RFC does
not intend to revisit its recommendations. However, there remain cases where
this drawback is highly undesirable or simply not acceptable:

> 1. Until we have implemented good editor integration, the developer experience

Reviewer: [question] I thought we accepted that editor integration (to include syntax highlighting etc) was a blocker to implementation (with the idea that it would drive implementation). When did that change?

   of using `.gjs`/`.gts` could be abysmal (without any syntax highlighting or
   completion at all, etc). Even after we ship good editor integration for the
   mainstream editors, this will likely remain the case for a subset of less
   used editors that we did not or could not prioritize supporting.

> 2. There may be editors and environments that are impossible for us to support,

Reviewer: This is not unique to this RFC but applies to Ember in general. As such, I am not sure it's relevant to include as a supporting reason.

   runnable code samples, generating components dynamically at runtime, etc.

From here on, we will refer to users and use cases collectively as "standard
JavaScript environments". Under these circumstances, users would be unable to
adopt the `<template>` feature, which means they cannot take advantage of the
numerous benefits unlocked by the feature, such as the strict mode opt-in.

Also, if an addon chose to only provide template constructs (e.g. helpers and
components) exclusively via importable modules (as opposed to merging them into
"App Javascript"), then there will be no easy way to access these template
constructs from those environments.

It is not technically impossible to accomplish these same goals in standard
JavaScript environments. After all, the `<template>` feature is specified using
primitives that already exist. Users in standard JavaScript environments can
just use those primitives directly.

However, that is not an ideal outcome. `<template>` is a user-facing feature,
and is poised to (or at least well-positioned to) become the main way Ember
users read, write and reason about components in the next edition of Ember when

> the feature is fully rolled out.

Reviewer 2: `<template>` vs `template()`.  What prevents the latter from wide adoption and use in the community?  Do we just expect users to prefer the format in rfc 779?

   around it (without using the `${ ... }` syntax), which can be confusing. It
   also means that it is impossible to provide a runtime implementation using
   the same syntax that works in environments that does not permit a build step.

4. It uses `static template = ` to associate templates to classes. This implies
   runtime behavior that is actually not true (e.g. TypeScript will believe the
   `template` property exists on the class).

5. Values consumed by the template but is otherwise unused in the rest of the
   file may generate errors/warnings from linters or language servers.

To address these issues, this RFC propose that we introduce a new `template()`
function that servers as a middle ground between the `<template>` language
extension and the low-level primitives.

- It is designed to have the same semantics as the `<template>` feature (such
  as the strict-mode opt-in, returning a template-only component when not
  attached to a class), providing the same high-level programming model for
  authoring components in standard JavaScript environments.

- It requires manually supplying the lexical scope variable bindings (the
  "scope function").

> - It uses a [static initializer block][static-block] to associate the template.

Reviewer 2: Based on reading the Class code snippets, the proposal is to use the static initializer block to side effect the association of the template?  Or is there actual static initialization on the class itself?  Perhaps inconsequential, but was hoping to enhance my understanding of this proposal.

  platforms where syntax highlighting is not yet available for the custom
  `.gjs` format (such as Discord and GitHub today), but where it's desirable to
  still retain the highlighting for the JavaScript/TypeScript portions.

## Detailed design

This RFC proposes to re-specify the `<template>` language extension into a
de-sugaring into `template()` calls:

1. Top-level declaration

   ```
   import Foo form 'somewhere';

   <template>
     <Foo />
   <template>
   ```

   becomes

   ```js
   import { template } from '@ember/template-compilation';
   import Foo form 'somewhere';

   export default template(`
     <Foo />
   `, () => ({ Foo }));
   ```

2. Expression

   ```
   import Foo form 'somewhere';

   export const Bar = <template><Foo /><template>;
   ```

   becomes

   ```js
   import { template } from '@ember/template-compilation';
   import Foo form 'somewhere';

   export const Bar = template(`<Foo />`, () => ({ Foo }));
   ```

3. Class

   ```
   import Foo form 'somewhere';

   export default class Bar {
     <template><Foo /><template>
   }
   ```

   becomes

   ```js
   import { template } from '@ember/template-compilation';
   import Foo form 'somewhere';

>    export default @template(`<Foo />`, () => ({ Foo }) class Bar {

Reviewer 3: can / should this RFC mention the pre-decorators version of class decorator usage?

I think it'd be:

```js
export default template(`<Foo />`, () => ({ Foo }))(class Bar {
});
```

Author: Yeah we can definitely bikeshed what is the most ergonomic version of this should be. I don't super love the decorator approach either, I just pushed a commit for a version using a static initializer block (matching the template/backtick version)

Reviewer 3: well, my motivation here is, I think aligned with the motivation of this RFC, to potentially allow shipping of ember "stuff" without transpilation.  Who knows how long it'll take before decorators are native, ya know?

Author: What do you think about the static initializer block version? I like that it keeps it inside the class body, matching `<template>` more closely. It always does not _intrinsically_ require the use of static initializer blocks if it's not supported in your environment, it's basically just a "three arguments version" of the function and you can wrap it around the class if you want to.

Author: I pushed an update making the static initializer block version the preferred one, and moved the class decorator to the alternatives section.

Reviewer 3: I have internal conflict. So like,

I'm used to having to do:
```js
setComponetnTemplate(hbs`my stuff`, MyClass);
```
which _could_ look like very similar to the decorator usage of template (which, maybe for other readers, class decorators are syntax sugar for functions that take a class and return a class)
```
@setComponentTemplate(hbs`my stuff`)
class MyClass {

}
```

<details><summary>implementation</summary>

```js
export function setComponentTemplate(...args) {
  if (args.length === 2) {
    _setComponentTemplate(...args);
    
    return args[1];
  }
  
  if (args.length === 1) {
    return (Klass) => {
      _setComponentTemplate(args[0], Klass);
      
      return Klass;
    }
  }
  
  throw new Error(`Unsupported arity: ${args.length}`);
}

```

</details>

Now, for the static initializer block -- I had never seen these in use before now, so that's exciting (TIL, etc) -- I can't _comfortably_ have an informed opinion on them yet though, only gut reactions.

And my gut reaction is that the static initialize block achieves the intent we want:
 - a component's template is not dynamic (or rather, you're in for a world of pain if you make it dynamic)
 - it's clear that the template is the same for every instance of the component
 - the association between template and component is also clear (whereas in can be missed with things like `setComponentTemplate`
 
 And what I'm wary of:
  - people unfamiliar with static initializers may be like "wtf? what is `this` in a static context", "why is JS weird?" (lol), "not `this` again" -- to which I say, _learn the language you're using, ask questions, etc,_
  - it's more typing / more blocks than just calling the function (but with the function, you have the same problem as `setComponentTemplate`, where the association is less obvious.
  
  I'm actually just realizing that we can use static initializers with `setComponentTemplate` today... which I think I may want to explore in my own stuff:
  
  ```js
  class Demo extends Component {
    static {
      setComponentTemplate(hbs`hello @name`, this);
    }
  }
  ```

So.... I think I'm in _favor_ of static initializers _because_ of the implied association between template and component.

   ```

3. Class

   ```
   import Foo form 'somewhere';

   export default class Bar {
     <template><Foo /><template>
   }
   ```

   becomes

   ```js
   import { template } from '@ember/template-compilation';
   import Foo form 'somewhere';

   export default class Bar {
     static { template`<Foo />` }
   }
   ```

**TBD**

## How we teach this

**TBD**: While the `template()` function is a bit cumbersome to use in practice
I think what it does is in fact quite easy to understand and teach. It may
actually help with teaching `<template>` if we can show what it desugars into.

## Drawbacks

**TBD**: Having the ability to supply the variable binding closure manually may
open it up to misuse (i.e. the closure does more than just simply capturing the
surrounding lexical scope, or does so in ways we didn't intend). However, this
is not a new problem in a sense, as the primitives already exists, and the
status quo is "if you cannot use `<template>`, use the primitives directly.
Also, with default helper manager, you can pretty much do anything along those
lines anyway with an inline closure that you invoke inside the template.

## Alternatives

**TBD**

## Unresolved questions

> **TBD**

Reviewer 3: would we have the ability to use this at runtime?

I imagine this `template` function _always_ converts to the underlying format, but in apps, a build step could swap out the implementation -- allowing `template` to exist at runtime, if imported and used dynamically, such as the use cases of:
 - https://github.com/NullVoxPopuli/ember-repl
 - [limber.glimdown.com/](https://limber.glimdown.com/)

Author: Yes that's the idea, see:

> It is designed to be pre-processed at build time where possible, just like the `<template>` feature and today's "official" `hbs` tag. However, in cases where this is not possible, there will be a runtime implementation available provided the template compiler is also available at runtime.

Author: Which is why it's important that the `template()` version doesn't do any magic that is not actually possible at runtime.

Reviewer 3: YAAAS

  tools to recognize the format, the tagged template literal form can be used,
  which automatically captures the lexical scope variable bindings. This
  directly replaces [ember-template-imports][ember-template-imports]'s "hbs
  backtick" format but brings its naming into alignment with the `<template>`
  feature, and uses a static initializer block to associate the template.

  This form will produce an error at runtime if not pre-processed out
  as a correct runtime implementation is impossible.

  Despite the drawbacks mentioned above, this format may still be preferred as
  a transitional tool for early adopters while editor integrations are being
  worked on. It may also be useful for communicating with code snippets in
  platforms where syntax highlighting is not yet available for the custom
  `.gjs` format (such as Discord and GitHub today), but where it's desirable to
  still retain the highlighting for the JavaScript/TypeScript portions.

## Detailed design

This RFC proposes to re-specify the `<template>` language extension into a
de-sugaring into `template()` calls:

1. Top-level declaration

   ```

>    import Foo form 'somewhere';

Reviewer 4: from, not form :)

   <template>
     <Foo />
   <template>
   ```

   becomes

   ```js
   import { template } from '@ember/template-compilation';

>    import Foo form 'somewhere';

Reviewer 4: from, not form :)

"inline {{let}}" so in practice it is not an issue. However, some polyfills for
template features also uses the AST plugin API and those polyfilled features
will not be available to templates compiled at runtime.

#### Opting In

Today, applications can opt-in to bundling the runtime compiler with something
along these lines in their `ember-cli-build.js`:

```js
app.import("node_modules/ember-source/dist/ember-template-compiler.js");
```

The exact details depend on the version of `ember-source` and is not very well
documented. It also does not align with the direction we are headed, where we
encourage using the modules system and imports to express dependencies. This
configuration is also not friendly to tree-shaking/code-splitting, as it forces
the template compiler to be included into the main/initial bundle, when it may
only be needed in one of the infrequently used routes.

This RFC takes the opportunity to improve on this by proposing a new module –
`@ember/template-compilation/runtime`. When this module is imported, either as
a "side-effect" import statement, or if its exported values are imported, then
it serves as an opt-in for enabling runtime template compilation.

> For the time being, we propose that this module should have a single named

Reviewer 3: as someone who's worked with runtime compilation a bit, it's _very_ useful to have all the utilities used to make `compileTemplate`: https://github.com/NullVoxPopuli/ember-repl/blob/main/addon/hbs.ts#L51

for example, building a component for block-based translations: https://github.com/RuslanZavacky/ember-intl-component/blob/try-upgrade/addon/components/i18n/index.js#L44

Reviewer 3: by limiting the re-exports, this re-blocks us from implementing component + block based translations.

cc @ruslanzavacky

Reviewer 3: clarification (less strong wording):
re-blocks:
_makes us continue to rely on :shrug:-compatible versions of `@glimmer/syntax`

Reviewer 3: `template` overall, makes this much nicer though:

```js
import { template } from '@ember/template-compilation/runtime';
import { getTemplateLocals } from '@glimmer/syntax'; // would prefer this be re-exported as well

// ...

  htmlComponent(i18nString) {
    i18nString = i18nString.replace(/\[\[\[(.*?)\]\]\]/g, '{{yield to="$1"}}');

    let owner = getOwner(this);
    let locals = getTemplateLocals(i18nString);
    let scope = {};

    for (let local of locals) {
      let kebabName = dasherize(local);
      let componentFactory = owner.factoryFor(`component:${kebabName}`);

      scope[local] = componentFactory.class;
    }

    return template(i18nString, () => scope); // no need for a while library for this step (bye ember-repl!)
}    
```

Author: I have a few questions, but does

1. `import { compile } from '@ember/template-compilation'` not work for you today for some reason? (If so not sure re-exporting that really helps?)
2. Can you not fill the same use cases with `template`/components? In what situations does it super matter that it returns a raw template as opposed to a template-only component?

Author: I think we resolved this in a side conversation @NullVoxPopuli? I don't think we actually know of any use cases where it is very important that `compile()` is used instead of `template()`, which is the only other export from that module.

export – `template`, which is a re-export of the `template()` function we
described above. It has the same semantics as `template()` from the main module
except it is always invalid to use this imported function as a tag, so linters
and TypeScript can trivial flag this invalid usage.

> The purpose of the re-exported `template()` from the runtime module is to

Reviewer 3: will we throw errors if the non-runtime module makes it into a build? (even if accidentally somehow, where a linter wouldn't be able to help?)

or, does it matter?

If the non-runtime module _happens_ to make it into the build, is it the exact same as the runtime module?

what are the drawbacks of only having one module instead of also having the runtime module?

Author: It won't. `template()` from the runtime module is just a re-export so yeah they are exactly the same. The reason to import from the main module (plus sticking with the statically analyzable subset, both of which you should do 95% of the time), is for your code to be agnostic to where the compilation happens. The reason to import from the runtime module is to make it clear that the template should never be processed at build time.

Reviewer 3: so is it just a hint to tooling? does it instead make sense to provide a way to analyze usage of `template` within a module to determine if it should be left in or compiled away?

Author: Sure. It's more than a guarantee than a hint, and it's not just to tools, it also makes it clear to whoever is reading the code that it is deliberate. You can look at the "idiomatic usages" section. The idea is we should implement all the same rules in the linter and warn you about non-idiomatic usages. And the instructions will tell you to either not do the thing that made it not statically analyzable, or to switch over to use the runtime import to make it clear that it's deliberate. However it's "just a lint". Even if you don't change the import and let it ship to the browser, it would still work (but you will probably get an additional warning).

Author: I think I cleared this up some more in the most recent update

---
Stage: Accepted
Start Date: 2022-04-16
Release Date: Unreleased
Release Versions:
  ember-source: vX.Y.Z
  ember-data: vX.Y.Z
Relevant Team(s): Ember.js
RFC PR: https://github.com/emberjs/rfcs/pull/813
---

# Layering and Desugaring for First-Class Component Templates

## Summary

This RFC propose to introduce a `template()` function as an alternative for the
`<template>` syntax extension defined in [RFC #779][rfc-779] using standard
JavaScript, for use in environments where the custom syntax extension is not
available, and also as a publishing format for addons on NPM:

[rfc-779]: ./0779-first-class-component-templates.md

```js
import { template } from "@ember/template-compilation";
import Hello from "my-app/components/hello";

> // <template><Hello></template> becomes...

Reviewer 5: Should't it be `<template><Hello /></template>` ?
If so, the typo is in the next block also.

Reviewer 3: yup

export default template(`<Hello />`, () => ({ Hello }));

// export const Foo = <template><Hello /></template> becomes...
export const Foo = template(`<Hello />`, () => ({ Hello }));

// export class Bar {
//   <template><Hello /></template>
// }
export class Bar {
  static {
    template(`<Hello />`, () => ({ Hello }), this);
  }
}
```

Idiomatic usages of the `template()` function can be pre-compiled by a build
step if desired, but a runtime implementation can also be used via an opt-in
described below.

We also propose an alternative form that uses the `template()` function as a
tag function for JavaScript tagged template strings. However, this variant does
not have a runtime implementation and must be handled by a build step:

```js
import { template } from "@ember/template-compilation";

> import Hello from "my-app/components/hello";

Reviewer 5: I wonder if this import wouldn't be detected as inactive by the syntax hyghlighter :thinking:

Author: I pushed some updates that clarified the motivations for the RFC. This is definitely not intended to _replace_ the `<template>` proposal.

Regarding your specific question, yeah, it is likely that without a lot of additional configurations/integrations the editor will probably not be able syntax highlight the content of the code block. Variables that are referenced only from within the template blocks will likely be (incorrectly) marked as unused.

These drawbacks are enumerated in RFC #779. Perhaps counterintuitively, the experience from the prototyping phase was that it would be _more_ difficult to properly support these things inside a `.js`/`.ts` file, because the `.js`/`.ts` language support plugins generally does not provide a way to do that for (what appears to be) arbitrary string literals. That's why the `<template>`+ custom file type approach was chosen instead. We'd have to do the work of wrapping/forking the JS/TS language support plugins _anyway_, and overall we believe it will be a better experience/more practical to have these Ember-specific plugins target Ember-specific file formats.

This RFC does not suggest that we revisit any of these, but just acknowledging that there are situations in which it is not feasible/possible to use/support the custom file format, and it will still be important to have _some_ (even if degraded) form of the `<template>` feature/programming model available in those environments.

  "scope function").

- It uses a [static initializer block][static-block] to associate the template.

  [static-block]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Class_static_initialization_blocks

- While it is more cumbersome to use and have a degraded experience (e.g.
  possibly lacking hbs syntax highlighting) compared to first-class component
  templates, it is still designed to be ergonomic to use, to the extent
  possible in standard JavaScript environments.

- It is designed to be pre-processed at build time where possible, just like
  the `<template>` feature and today's "official" `hbs` tag. However, in cases
  where this is not possible, there will be a runtime implementation available
  provided the template compiler is also available at runtime.

- For environments where it is possible to configure the development and build
  tools to recognize the format, the tagged template literal form can be used,
  which automatically captures the lexical scope variable bindings. This
  directly replaces [ember-template-imports][ember-template-imports]'s "hbs
  backtick" format but brings its naming into alignment with the `<template>`

>   feature.

Reviewer 6: What would be an environment which can handle *that* preprocessing step but which *couldn't* handle `<template>`?

of RFC #779 which references its compilation output, and where it specified the
feature in terms of the low-level primitives such as `precompileTemplate()`. If
this RFC is accepted, the `<template>` feature should only be understood in
terms of `template()`, which will be a stable and public API that stands on its
own.

Any build-time optimization that pre-compiles the `template()` calls at build
time are required to ensure their output to have equivalent semantics as the
runtime `template()` calls, assuming the same set of AST plugins are applied.
Their exact compilation output should be considered private implementation
details which may change over time and should not be relied upon.

Once this RFC is accepted, any future changes (including deprecations) to the
primitive APIs referenced in RFC #779 should have no direct bearing on the
`<template>` or `template()` features.

### Tagged Template Literals

Alternatively, the `template()` function can also be used as a tag function for
a JavaScript tagged template string. This alternate form automatically captures
the lexical scope variable bindings. However, this form does not have a runtime

> implementation and must be processed by a build step.

Reviewer 6: This *does not* actually "work" in any of the target environments described above. For example, in CodePen, the code would parse, but you wouldn't be able to *use* it at all: The template part of it would be completely inert! I’m curious what you see this providing for users?

compilation and will likely remain to be the case for the foreseeable future,
as these plugins are often authored or packaged in ways that are inherently not
browser-friendly. This usually only used for optional syntax extensions such as
"inline {{let}}" so in practice it is not an issue. However, some polyfills for
template features also uses the AST plugin API and those polyfilled features
will not be available to templates compiled at runtime.

#### Opting In

Today, applications can opt-in to bundling the runtime compiler with something
along these lines in their `ember-cli-build.js`:

```js
app.import("node_modules/ember-source/dist/ember-template-compiler.js");
```

The exact details depend on the version of `ember-source` and is not very well
documented. It also does not align with the direction we are headed, where we
encourage using the modules system and imports to express dependencies. This
configuration is also not friendly to tree-shaking/code-splitting, as it forces
the template compiler to be included into the main/initial bundle, when it may
only be needed in one of the infrequently used routes.

> This RFC takes the opportunity to improve on this by proposing a new module –
> `@ember/template-compilation/runtime`. When this module is imported, either as
> a "side-effect" import statement, or if its exported values are imported, then
> it serves as an opt-in for enabling runtime template compilation.

Reviewer 6: ```suggestion
This RFC takes the opportunity to improve on this by proposing a new module –
`@ember/template-compilation/runtime`. We will not compile out the `template()`
function imported from this module. Additionally, importing the module as a
"side-effect" import statement will also serve as an opt-in for including it for
dynamic runtime template compilation.
```

(I tripped over the original phrasing here and am suggesting this after chatting about it with @chancancode in Discord.)

of usages to guarantee that is the case, and this re-export from the runtime
module offers a way to signal that the deviations are deliberate.

This RFC does not propose to add re-exports for other features found in the
`@ember/template-compilation` package. Conceptually, it makes sense to provide
the same ex-export for at least the `compile()` function. However, as we move
to a world where components are used everywhere and the need for "standalone
templates" dwindles, it is unclear that the `compile()` function is needed in
the long run. In any case, if that turns out to be necessary, a future RFC can
propose to add support for that and the precedence set forth in this RFC should
make that a straightforward process.

### TypeScript

We don't expect this feature to pose any significant challenges in terms of
TypeScript support, other than having a few overloaded signatures/purpose on
the same `template()` function may make things a little cumbersome, but not
prohibitively so. The signature of the function can be described as such:

```ts
// These types are defined in other RFCs
interface Signature {
  /* ... */
}

> interface Component<S extends Signature> {

Reviewer 6: Note: the `extends` thing actually causes a bunch of annoying issues in terms of what it does/n't support (for good reasons around interface merging) and also doesn't actually add any value (because they're type-level extensions so TS will happily let you typo away: `Arsg` is a totally legitimate thing to extend `Signature` with!), so:

```suggestion
interface Component<S> {
```

  /* ... */
}

type Scope = () => Record<string, unknown>;

> function template<S extends Signature>(source: string): Component<S>;
> function template<S extends Signature>(

Reviewer 6: ```suggestion
function template<S>(source: string): Component<S>;
function template<S>(
```

  source: string,
  scope: Scope
): Component<S>;

> function template<S extends Signature, C extends Component<S>>(

Reviewer 6: ```suggestion
function template<S, C extends Component<S>>(
```

  source: string,
  scope: Scope,
  target: C
): C;

// For the build-time tagged template literal extension. By not specifying
// additional arguments here, TypeScript actually enforces that the tagged
// string literal cannot have any interpolation expressions.

> function template<S extends Signature>(

Reviewer 6: ```suggestion
function template<S>(
```

  source: TemplateStringsArray
): Component<S>;
```

One nice thing about this design is that there is a natural place for the type
(generic) argument for the template's signature, negating the need for the

> special `<template[Signature]>` syntax.

Reviewer 6: This actually does *not* solve the issue from a general authoring perspective, though, I think? After all, we're not going to be teaching people to write it this way in general, and we’re *definitely* not going to be implementing Glint support for it.

Author: Yep, it should probably say it gives you a place to emit into. But what I was trying to say (badly) is that it does not require specifically designing something in this API that does that equivalent of `<template[Signature]>` because it already naturally exists

Reviewer 6: Ah, yep, that’s clearer and I agree: We still need surface syntax for it, but it’s nice that there’s a natural home for it in the compile target. 👍🏼

## How we teach this

The teaching recommendations in [RFC #779][rfc-779] shall remain in full force.
We continue to recommend teaching materials to be focused around `<template>`
and `.gjs`. The `template()` feature is recommended only in standard JavaScript
environments where adoption of `<template>` and `.gjs` is not feasible.

Because `template()` is a standard JavaScript API that has a "real" import
location, there is a natural place to document and describe its behavior in the
API docs. This should be the primary reference of the feature, and should also

> where we document the idiomatic subset as well.

Reviewer 6: We should use that as a hook for describing the whole end-to-end bits in the API docs, including `<template>`, too, since as you note below it's the actual desugaring for that.

whitespace just to have the browser fold them away when rendered into the DOM
is a not insignificant amount of wasted space in app bundles today. It also
just creates unnecessary work for the browser and the rendering layer.

On the other hand, there is the argument that whitespaces _can_ be significant.
Sometimes, this can be observed from uses of the `<pre>` element, but really,
the whitespace handling for any element (including `<pre>`) can be changed
using the CSS `white-space` property, so it is not necessarily safe to assume
that we can fold/collapse whitespaces in templates just because that's the
default behavior of the browser.

However, even with that in mind, the default conclusion for the whitespace
handling in `<template>` tags is _still_ not useful, _especially_ when the
author is trying to preserve significant whitespace for something like code
samples. This is because `<template>` tags often appears in JavaScript
contexts that already has leading indentation (such as inside a class body), so
in practice, if we preserve the whitespace exactly as it appears inside the
`<template>` tag, it still would not do what the author was trying to

> accomplish.

Reviewer 6: Strongly agree, and in line with the discussions we've had offline, I think we should probably apply the `HEREDOC`/etc. rules: *leading* indentation up to the indentation of the line on which the opening `<template>` appears should be stripped.

// export class Bar {
//   <template><Hello /></template>
// }
export class Bar {

>   static {

Reviewer 3: I was initially against this, but I worked through it here
https://github.com/emberjs/rfcs/pull/813#discussion_r851770215

and now I'm hugely in favor :100:
